util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,249 @@
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def get_resource_local_filepath(
+    resource_id: str, resource_version: str, base_output_path: TextIO
+) -> Path:
+    """
+    This function returns the local filepath for a resource
+    :param resource_url: URL field of the resource
+    :param base_output_path: Base output directory absolute path
+    :return: Local filepath for the resource
+    """
+    filename = f"{resource_id}-{resource_version}"
+
+    filepath = Path(base_output_path).joinpath(filename)
+
+    return filepath
+
+
+def download_resources(resources: List[Dict], output: TextIO):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filepath = get_resource_local_filepath(
+                resource["id"], resource["resource_version"], path
+            )
+
+            if not filepath.exists() or (  # If the file does not exist
+                "url-md5sum"
+                in resource.keys()  # If it exist but the md5sum is wrong.
+                and hashlib.md5(filepath.read_bytes()).hexdigest()
+                != resource["url-md5sum"]
+            ):
+                print(f"Downloading {url} to {filepath}")
+                request.urlretrieve(url, filepath)
+            resource["url"] = filepath.absolute().as_uri()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Get resources from the database"
+    )
+
+    parser.add_argument(
+        "--config-file-path",
+        type=str,
+        default=Path(__file__)
+        .resolve()
+        .parent.joinpath("gem5_default_config.py"),
+        help="Filepath to the gem5 config file",
+    )
+
+    parser.add_argument(
+        "--output_dir",
+        type=str,
+        default=Path.cwd(),
+        help="Output directory path, default is the cwd."
+        "The resources.json and all resources will be saved in this directory",
+    )
+    args = parser.parse_args()
+
+    output_path = Path(args.output_dir)
+
+    # Get the gem5 config file from the  file
+
+    with open(args.config_file_path) as f:
+        gem5_config_file_contents = f.read()
+    gem5_config_str = gem5_config_file_contents.split("=")[-1]
+    gem5_config_json = eval(gem5_config_str)
+    gem5_config_gem5_resources = gem5_config_json["sources"]["gem5-resources"]
```

Comment:
If we're using the Python file to contain this data. Why don't we just `from .gem5_default_config import config`?

((Though I am advocating for this to be a JSON file rather than a Python file)).

Reply:
If we were to keep this as a Python file _I think_ there's a way to dynamically import the file.

Though this code is massively simplified if we just have the file in a JSON format.